File: sorting/mergesort/python/mergeSort.py

def Sort(input : list)->list:
    if len(input)<=1:
        return input
    midPoint = len(input)//2
    left = Sort(input[:midPoint])
    right = Sort(input[midPoint:])
    return merge(left,right)

def merge(left:list,right:list)->list:
    merged = []
    i=0
    j=0
    while i<len(left) and j<len(right):
        if left[i]<right[j]:
            merged.append(left[i])
            i+=1
            continue
        if left[i]>right[j]:
            merged.append(right[j])
            j+=1
            continue
        if left[i]==right[j]:
            merged.append(left[i])
            merged.append(right[j])
            i+=1
            j+=1
            continue

    if i<len(left):
        merged.extend(left[i:])
    if j<len(right):
        merged.extend(right[j:])
    return merged

File: sorting/mergesort/javascript/mergeSort.js

function Sort(input)
{
  if(input.length<=1)
  {
    return input;
  }
  let midPoint = input/2;
  left = Sort(input.slice(0,midPoint));
  right = Sort(input.slice(midPoint));
  return merge(left,right);
}

function merge(left,right)
{
  let merged = []
  for(let i=0,j=0;i<left.length&&j<right.length;)
  {
    if(left[i]<right[j])
    {
      merged.push(left[i]);
      i++;
      continue;
    }
    if(left[i]>right[j])
    {
      merged.push(right[j])
      j++;
      continue;
    }
    if(left[i]==right[j])
    {
      merged.push(left[i]);
      merged.push(right[j]);
      i++;j++;
      continue;
    }
  }
  if(i<left.length)
  {
    merged.concat(left.slice(i));
  }
  if(j<right.length)
  {
    merged.concat(right.slice(j));
  }

  return merged;
}
